_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+
+### Akcje
+Potrzebujemy teraz możliwości wywoływania akcji, żeby powodować zmiany w store. Aplikacja będzie przewidywała dwa zdarzenia: wystosowanie żądania do API po pomysł na jakieś zajęcie oraz dodanie aktywności do listy. Wygląda to następująco:
+
+```typescript
+export const addActivityType = 'Add activity';
+export const getActivityType = 'Get activity';
+
+export const addActivity = createAction(addActivityType, props<ActivityItemModel>());
+export const getActivity = createAction(getActivityType);
+```
+
+Każda akcja musi mieć zdefiniowany swój typ. NgRx używa do tego tzw. _literal type_, czyli po prostu używa typu string do identyfikacji akcji (u nas odpowiednio `Add activity` oraz `Get activity`. W nazywaniu akcji mamy pełną dowolność. Dodatkowo w przypadku dodawania nowej aktywności do listy musimy ją przekazać w ciele akcji przy pomocy metody `props<>()`. Dzięki wykorzystaniu funkcji `createAction` z biblioteki NgRx tworzenie akcji, jak widać, jest bardzo proste.
+
+### Efekt
+Jak wspomniałem, będziemy korzystać z zewnętrznego API. W tym celu napiszemy efekt, który będzie reagował na akcję `Get activity`, następnie wystosowywał żądanie HTTP, a po otrzymaniu odpowiedzi wywoływał akcję `Add Activity`.
+
+```typescript
+@Injectable()
+export class ActivityEffect {
+    constructor(private actions$: Actions,
+                private http: HttpClient) {
+    }
+
+  getActivity$ = createEffect(() => this.actions$.pipe(
+    ofType(getActivityType),
+    mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity')
+      .pipe(
+        map(response => (addActivity(response))),
+        catchError(() => EMPTY)
+      ))
+  ));
+}
+```
+
+Pojawia się tutaj dość dużo zagadnień, zatem spójrzmy na ten fragment kodu od góry do dołu. Najpierw informujemy Angulara, że efekt to tak naprawdę wstrzykiwalna klasa, podobnie jak zwykłe serwisy. W konstruktorze zapewniamy dostęp do serwisu HTTP oraz do zmiennej typu `Actions`. Jest to w gruncie rzeczy serwis udostępniający wszystkie wywołane akcje od momentu ostatniego uruchomienia reducera (czyli dzięki temu damy radę reagować na przychodzące akcje). 
+
+Następnie definiujemy sam efekt:
+* zaglądamy do strumienia akcji 
+```typescript 
+this.actions$.pipe()
+```
+* szukamy tylko akcji określonego typu 
+```typescript
+ofType(getActivityType)
+```
+* reagujemy na każdą z nich poprzez wywołanie żądania HTTP i zaglądamy do strumienia zawierającego odpowiedź z API 
+```typescript 
+mergeMap(() => this.http.get<ActivityItemModel>('https://www.boredapi.com/api/activity').pipe()
+```
+* mapujemy odpowiedź na wywołanie akcji `map(response => (addActivity(response)))`
+```typescript 
+map(response => (addActivity(response)))
+```
+* w przypadku błędu zwracamy pustą wartość Obervable
+```typescript 
+catchError(() => EMPTY)
+```
+
+W wyniku działania tego efektu każde wywołanie akcji `Get Activity` przy poprawnej odpowiedzi z API wywoła akcję `Add Activity` z otrzymaną czynnością.
```

Comment:
może warto wspomnieć o innych przypadkach efektów poza http? np. gdy mamy do zbudowania akcje warunkowo, z nietrywialną logiką i wołane z wielu miejsc to czasem mamy efekt generujący takie akcje, które później wpływają na reducery i inne efekty. wtedy nie ma http, jest np. złożona obliczeniowo lub logicznie operacja. ale może to tylko my coś źle robimy? :)

Reply:
Można wspomnieć, natomiast nie chcę zawalać wpisu przykładami i „o patrz, tak tez możesz zrobić!”, bo to jednak wprowadzenie do NgRx, a nie kompleksowy poradnik